### ISSUE TITLE
Feature request: let tf.contrib.layers.batch_norm normalize over more axes than just the last one

### ISSUE THREAD

**Issue by carlthome** [tensorflow/tensorflow#7091]
When using `tf.nn.batch_normalization` and `tf.nn.moments` it's possible to choose which axes are aggregated and thus which dimensions are normalized over. For `tf.contrib.layers.batch_norm` and `tf.contrib.layers.layer_norm` the assumption is that only the last dimension will be normalized over. This is standard practice for filters in ConvNets, or the features in fully-connected layers, but there are use cases where it would be nice to independently normalize across both width and filters (for example audio spectrograms, where "width" could be seen as the time axis). Could the layers API be changed to be a little more flexible or is the current consensus that `tf.nn.batch_normalization` is the way to go?

**Comment by yaroslavvb** [comment]
It seems useful, ie, there's [Layer Normalization](https://arxiv.org/abs/1607.06450) which got good performance which is transpose of batch normalization, and [Weight Normalization](https://arxiv.org/abs/1602.07868)

 I wonder if it should be called something else though, since `batch` in `batch_norm` refers to batch dimension (0'th dimension).

I don't see a way to specify axis in `tf.nb.batch_normalization`

`tf.nn.batch_normalization(x, mean, variance, offset, scale, variance_epsilon, name=None)`

**Comment by carlthome** [comment]
According to its docstring `tf.nn.batch_normalization` infers the axes from the input tensors (I assume `x', `mean`, `variance` etc. and that they're expected to match), so it didn't really need an `axis` parameter, I guess?

By the way, my feature request applies to the layer norm layer too. That too is useful to be able to apply to several dimensions instead of just the last, particularly for RNNs where batch norm is... tricky.

**Comment by yaroslavvb** [comment]
Oh, good point, it looks like it can normalize over arbitrary axis, taking specification from the missing dimension in `mean`/`variance` tensors

**Comment by drpngx** [comment]
What is the problem with using `tf.nn.batch_normalization`?

**Comment by carlthome** [comment]
`tf.nn.batch_normalization` is more low-level. Unlike the `layers` version it doesn't handle estimation of population statistics as decaying moving averages during training; it doesn't make a distinction between test time and train time statistics on its own; it requires you to manually introduce trainable variables for beta and gamma, and so on. It's rather tricky to setup, in fact, as it's basically just the mathematical operation and nothing more.

**Comment by drpngx** [comment]
Could you outline what the API change would look like?

**Comment by carlthome** [comment]
I'd expect `tf.contrib.layers.batch_norm(..., axis=[0, 1, 2])` to work, just it it does for `tf.nn.moments`.

For backwards compatibility (where all but the last dimension should be reduced) `axis=None` is necessary as the default argument because the shape of the input tensor is not known, with documented behavior that it aggregates every dimension except for the last.

**Comment by carlthome** [comment]
Or preferably a new normalization layer, that takes `mode` and `axis` so we could choose layer norm, batch norm, weight norm and not consider the particular implementation details of the competing normalization methods (they have the same purpose of minimizing covariate shift AFAIK, but with different tradeoffs in code complexity / normalization "quality"):

```python
tf.contrib.layers.normalize(..., mode='batch', axis=None)
tf.contrib.layers.normalize(..., mode='layer', axis=[0, 1, 2])
tf.contrib.layers.normalize(..., mode='weight', axis=[0, 1])
```

**Comment by martinwicke** [comment]
@fchollet what do you think about this?

**Comment by fchollet** [comment]
> For backwards compatibility (where all but the last dimension should be reduced) axis=None is necessary as the default argument because the shape of the input tensor is not known

I don't understand this part? What does this mean?

The default value for `axis` should be `-1`, which is the feature axis is essentially every situation except convolution layers in `data_format` NCHW.

I had not heard before that one would want to normalize on more than the features axis. But we can consider the API change if it seems worth it.

**Comment by carlthome** [comment]
> What does this mean?

Say tensors are `[batch_size, frames, height, width, channels]` (video). Then `axis=None` could assume the user want means/variances as `[batch_size, frames * height * width * channels]` and not `[batch_size, frames, height, width]` (`axis=-1`).

Uh, I'm confusing myself. :koala: 

> I had not heard before that one would want to normalize on more than the features axis. But we can consider the API change if it seems worth it. By which I mean we can consider allowing a tuple of axes as the axis argument: the axes to normalize over, e.g. [-2, -1] or [1, 2].

A common use case could be RNNs for audio and video, where you'd like to normalize whole sequences.

**Comment by martinwicke** [comment]
This would be a good feature to add to the existing batch_norm (in contrib).
